packages/core/src/rules/common/assertions/index.ts

```diff
@@ -0,0 +1,60 @@
+import { asserts, runOnKeysSet, runOnValuesSet } from './asserts';
+import { Assert, buildSubjectVisitor, buildVisitorObject } from './utils';
+import { Oas2Rule, Oas3Rule } from '../../../visitors';
+
+export const Assertions: Oas3Rule | Oas2Rule = (opts: object) => {
+  let visitors: any[] = [];
+
+  // As 'Assertions' has an array of asserts,
+  // that array spreads into an 'opts' object on init rules phase here
+  // https://github.com/Redocly/openapi-cli/blob/master/packages/core/src/config/config.ts#L311
+  // that is why we need to iterate through 'opts' values;
+  // before - filter only object 'opts' values
+  const assertions: any[] = Object.values(opts)
+    .filter((opt: unknown) => typeof opt === 'object' && opt !== null);
+
+  for (const [index, assertion] of assertions.entries()) {
+    const assertId = assertion.assertionId && `${assertion.assertionId} assertion` || `assertion #${index+1}`;
+
+    if (!assertion.subject) {
+      throw new Error(`${assertId}: 'subject' is required`);
+    }
+
+    const subjects: string[] = Array.isArray(assertion.subject) ? assertion.subject : [assertion.subject];
+
+    const assertsToApply: Assert[] =
+      Object.keys(asserts)
+        .filter((assertName: string) => assertion[assertName] !== undefined)
+        .map((assertName: string) => {
+          return {
+            assertId,
+            name: assertName,
+            conditions: assertion[assertName],
+            message: assertion.message,
+            severity: assertion.severity || 'error',
+            suggest: assertion.suggest || [],
+            runsOnKeys: runOnKeysSet.has(assertName),
+            runsOnValues: runOnValuesSet.has(assertName)
+          }
+        });
+
+    const shouldRunOnKeys: Assert | undefined = assertsToApply.find((assert: Assert) => assert.runsOnKeys && !assert.runsOnValues);
+    const shouldRunOnValues: Assert | undefined = assertsToApply.find((assert: Assert) => assert.runsOnValues && !assert.runsOnKeys);
+
+    if (shouldRunOnValues && !assertion.property) {
+      throw new Error(`${shouldRunOnValues.name} can't be used on all keys. Please provide a single property.`);
+    }
+
+    if (shouldRunOnKeys && assertion.property) {
+      throw new Error(`${shouldRunOnKeys.name} can't be used on a single property. Please use 'property'.`);
+    }
+
+    for (const subject of subjects) {
+      const subjectVisitor = buildSubjectVisitor(assertion.property, assertsToApply, assertion.context);
+      const visitorObject = buildVisitorObject(subject, assertion.context, subjectVisitor);
+      visitors.push(visitorObject);
+    }
+  }
```

Comment:
What the benefit of avoiding mutation here?
Then it should be something like that:
```suggestion
    visitors = [
       ...visitors,
       ...subjects.map(subject => {
            const subjectVisitor = buildSubjectVisitor(assertion.property, assertsToApply, assertion.context);
            return buildVisitorObject(subject, assertion.context, subjectVisitor);
       })
    ]
```

Reply:
1. It always easier to keep track on your code if your variables don't change. 
2. In this case you don't need to spread `...visitors` since you there's no value assigned before unless I'm missing something. The only usage of `visitors` I see is here: https://github.com/Redocly/openapi-cli/pull/489/files?file-filters%5B%5D=.ts&show-viewed-files=true#diff-d9c7573397e9ca9187db6db4319df03e40bb5ec9a3eb067c973b0c225167978bR6

Reply:
The `visitors` array mutates through the 2 cycles, and as result we want to see a flat array of visitors.
Just discussed with you on the call 😃